zerver/migrations/0375_invalid_character_in_stream_name.py

```diff
@@ -0,0 +1,48 @@
+import unicodedata
+
+from django.db import migrations
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+# There are 66 Unicode non-characters; see
+# https://www.unicode.org/faq/private_use.html#nonchar4
+unicode_non_chars = set(
+    chr(x)
+    for x in list(range(0xFDD0, 0xFDF0))  # FDD0 through FDEF, inclusive
+    + list(range(0xFFFE, 0x110000, 0x10000))  # 0xFFFE, 0x1FFFE, ... 0x10FFFE inclusive
+    + list(range(0xFFFF, 0x110000, 0x10000))  # 0xFFFF, 0x1FFFF, ... 0x10FFFF inclusive
+)
+
+
+def character_is_printable(character: str) -> bool:
+    return not (unicodedata.category(character) in ["Cc", "Cs"] or character in unicode_non_chars)
+
+
+def fix_stream_name(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Stream = apps.get_model("zerver", "Stream")
+
+    streams = Stream.objects.all()
+    for stream in streams:
+        fixed_stream_name = "".join(
+            character for character in stream.name if character_is_printable(character)
+        )
+
+        if fixed_stream_name == stream.name:
+            continue
+
+        if fixed_stream_name == "":
+            stream.name = "Unknown stream {}".format(stream.id)
+            stream.save()
+            continue
+
+        similar_stream_name_count = Stream.objects.filter(name=fixed_stream_name).count()
+        if similar_stream_name_count > 1:
```

Comment:
Oh, and being >0 is a problem, not >1 -- we've not done the rename yet, so we expect this to be either 0, or 1.

Reply:
@alexmv I have used `.count`, which returns the total number of the object with the `name=fixed_stream_name`, which will include the stream which we are currently checking. I think `similar_stream_name_count` will be always `>=1`

Reply:
No, because we've not changed the `stream` object or called `stream.save`.  And because we have a unique index on `(realm_id, upper(name::text))`, we wouldn't even be able to _save_ the stream object if it was non-unique.

Reply:
Ok, got it. I have changed that.